### ISSUE TITLE
[Feature Request] Janus Generation Finetuning

### ISSUE THREAD

**Issue by eslambakr** [PKU-Alignment/align-anything#179]
### Required prerequisites

- [x] I have searched the [Issue Tracker](https://github.com/PKU-Alignment/align-anything/issues) and [Discussions](https://github.com/PKU-Alignment/align-anything/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [x] Consider asking first in a [Discussion](https://github.com/PKU-Alignment/align-anything/discussions/new).

### Motivation

Janus Generation Finetuning

### Solution

Hello everyone,

Thank u so much for your great efforts!

I wonder will you in the near future have any plan to support janus-pro finetuning for the generation part?

Thanks in advance!

**Comment by htlou** [comment]
Thanks for your interest! We already support fine-tuning the generation component in `scripts/janus/janus_sft.sh`. The process involves two steps:

1. First, tokenize your dataset using `projects/janus/supervised_tokenize.sh`
2. Then run `scripts/janus/janus_sft.sh` to perform the actual fine-tuning

You'll find detailed examples for both steps in our readme and tutorial. If you encounter any issues during this process, please feel free to return to this issue thread or open a new issue with your questions. We appreciate your contributions!

**Comment by david9934** [comment]
> Thanks for your interest! We already support fine-tuning the generation component in `scripts/janus/janus_sft.sh`. The process involves two steps:
> 
> 1. First, tokenize your dataset using `projects/janus/supervised_tokenize.sh`
> 2. Then run `scripts/janus/janus_sft.sh` to perform the actual fine-tuning
> 
> You'll find detailed examples for both steps in our readme and tutorial. If you encounter any issues during this process, please feel free to return to this issue thread or open a new issue with your questions. We appreciate your contributions!

Thank you for your response. From my understanding, the logits computed by Janus for image generation involve classifier-free guidance (CFG), i.e., `logits = logit_uncond + cfg_weight * (logit_cond - logit_uncond)`.

However, in align-anything, I observed that the loss calculation is as follows:

```python
def loss(self, sft_batch: SupervisedBatch) -> dict[str, torch.Tensor]:
    """Loss function for supervised finetuning."""
    outputs = self.model.forward(**sft_batch, task=sft_batch['task'])
    return {
        'loss': outputs.loss,
    }
```

It seems that there is no involvement of CFG in this calculation. 

Could it be that I missed some related details, or is it that image generation fine-tuning of Janus does not need to consider CFG?

Thanks for your kind help!

**Comment by htlou** [comment]
Thanks for asking! 

For the sake of code consistency, we didn't implement this loss in THIS repo directly. Instead, it's implemented in the `htlou/Align_Anything_Janus` repository(`janus/models/modeling_vlm.py`, lines 272-308), which would be installed following the installation process outlined in `projects/janus/README.md`. This loss will be activated during Janus training in `align-anything`.

If you encounter any issues during implementation, or your concern is not fully eliminated, please don't hesitate to reach out. We're happy to provide further assistance!

**Comment by david9934** [comment]
Thanks for your great work and timely response! 
My issue has been completely solved. It would help a lot!
